File: frontend/src/api/backtest.ts
import client from './client'

export interface BacktestCreate {
  name: string
  product_id: string
  strategy_type: 'dca' | 'grid'
  personality?: 'conservative' | 'moderate' | 'aggressive' | 'degen'
  config: Record<string, any>
  period_days: number
  starting_capital: number
}

export interface BacktestRun {
  id: string
  name: string
  product_id: string
  strategy_type: string
  personality: string | null
  config: Record<string, any>
  period_days: number
  start_date: string
  end_date: string
  starting_capital: string
  status: 'pending' | 'running' | 'completed' | 'failed'
  error_message: string | null
  ending_capital: string | null
  pnl: string | null
  pnl_pct: string | null
  max_drawdown_pct: string | null
  win_rate_pct: string | null
  sharpe_ratio: string | null
  total_trades: number | null
  created_at: string
  completed_at: string | null
  equity_curve?: { ts: number; equity: number; price: number }[]
  trades_log?: Array<Record<string, any>>
}

export interface BacktestQuota {
  tier: string
  max_period_days: number
  monthly_limit: number | null
  used_this_month: number
  remaining: number | null
}

export interface BacktestCatalog {
  products: string[]
  periods: number[]
  tiers: Record<string, { max_period_days: number; monthly_limit: number | null }>
}

export const getCatalog = () => client.get<BacktestCatalog>('/backtest/catalog')
export const getQuota = () => client.get<BacktestQuota>('/backtest/quota')
export const listBacktests = () => client.get<BacktestRun[]>('/backtest')
export const getBacktest = (id: string) => client.get<BacktestRun>(`/backtest/${id}`)
export const createBacktest = (data: BacktestCreate) =>
  client.post<BacktestRun>('/backtest', data)
export const deleteBacktest = (id: string) => client.delete(`/backtest/${id}`)

File: backend/app/schemas/backtest.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field


class BacktestCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)
    product_id: str = Field(..., description="e.g. BTC-USD")
    strategy_type: str = Field(..., pattern="^(dca|grid)$")
    personality: str | None = Field(default=None, pattern="^(conservative|moderate|aggressive|degen)$")
    config: dict[str, Any]
    period_days: int = Field(..., ge=7, le=1825)
    starting_capital: float = Field(..., gt=0)


class BacktestResponse(BaseModel):
    id: uuid.UUID
    name: str
    product_id: str
    strategy_type: str
    personality: str | None
    config: dict[str, Any]
    period_days: int
    start_date: datetime
    end_date: datetime
    starting_capital: Decimal
    status: str
    error_message: str | None = None
    ending_capital: Decimal | None = None
    pnl: Decimal | None = None
    pnl_pct: Decimal | None = None
    max_drawdown_pct: Decimal | None = None
    win_rate_pct: Decimal | None = None
    sharpe_ratio: Decimal | None = None
    total_trades: int | None = None
    created_at: datetime
    completed_at: datetime | None = None

    class Config:
        from_attributes = True


class BacktestDetailResponse(BacktestResponse):
    equity_curve: list[dict[str, Any]] | None = None
    trades_log: list[dict[str, Any]] | None = None


class BacktestQuotaResponse(BaseModel):
    tier: str
    max_period_days: int
    monthly_limit: int | None  # None = unlimited
    used_this_month: int
    remaining: int | None  # None = unlimited
